src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1504,133 +1612,136 @@ private async ValueTask<(Socket? socket, Stream stream)> EstablishSocksTunnel(Ht
-        private TaskCompletionSourceWithCancellation<HttpConnection?> EnqueueWaiter()
+        private void HandleHttp11ConnectionFailure(Exception e)
         {
-            Debug.Assert(Monitor.IsEntered(SyncObj));
-            Debug.Assert(Settings._maxConnectionsPerServer != int.MaxValue);
-            Debug.Assert(_idleConnections.Count == 0, $"With {_idleConnections.Count} idle connections, we shouldn't have a waiter.");
+            if (NetEventSource.Log.IsEnabled()) Trace("HTTP/1.1 connection failed");
 
-            if (_waiters == null)
+            lock (SyncObj)
             {
-                _waiters = new Queue<TaskCompletionSourceWithCancellation<HttpConnection?>>();
-            }
+                Debug.Assert(_associatedHttp11ConnectionCount > 0);
+                Debug.Assert(_pendingHttp11ConnectionCount > 0);
 
-            var waiter = new TaskCompletionSourceWithCancellation<HttpConnection?>();
-            _waiters.Enqueue(waiter);
-            return waiter;
-        }
-
-        private void IncrementConnectionCountNoLock()
-        {
-            Debug.Assert(Monitor.IsEntered(SyncObj), $"Expected to be holding {nameof(SyncObj)}");
+                _associatedHttp11ConnectionCount--;
+                _pendingHttp11ConnectionCount--;
 
-            if (NetEventSource.Log.IsEnabled()) Trace(null);
-            _usedSinceLastCleanup = true;
+                // Fail the next queued request (if any) with this error.
+                _http11RequestQueue.TryFailNextRequest(e);
 
-            Debug.Assert(
-                _associatedConnectionCount >= 0 && _associatedConnectionCount < _maxConnections,
-                $"Expected 0 <= {_associatedConnectionCount} < {_maxConnections}");
-            _associatedConnectionCount++;
+                CheckForHttp11ConnectionInjection();
+            }
         }
 
-        internal void IncrementConnectionCount()
+        private void HandleHttp2ConnectionFailure(Exception e)
         {
+            if (NetEventSource.Log.IsEnabled()) Trace("HTTP2 connection failed");
+
             lock (SyncObj)
             {
-                IncrementConnectionCountNoLock();
-            }
-        }
+                Debug.Assert(_associatedHttp2ConnectionCount > 0);
+                Debug.Assert(_pendingHttp2Connection);
 
-        private bool TransferConnection(HttpConnection? connection)
-        {
-            Debug.Assert(Monitor.IsEntered(SyncObj));
+                _associatedHttp2ConnectionCount--;
+                _pendingHttp2Connection = false;
 
-            if (_waiters == null)
-            {
-                return false;
-            }
+                // Fail the next queued request (if any) with this error.
+                _http2RequestQueue.TryFailNextRequest(e);
 
-            Debug.Assert(_maxConnections != int.MaxValue, "_waiters queue is allocated but no connection limit is set??");
+                CheckForHttp2ConnectionInjection();
```

Comment:
There is a cyclic call chain `CheckForHttp2ConnectionInjection` -> `AddHttp2ConnectionAsync` -> `HandleHttp2ConnectionFailure` -> `CheckForHttp2ConnectionInjection` which concerns me because it looks vulnerable to subtle bugs we can lead to going into an infinite loop.

Reply:
AddHttp2ConnectionAsync is never called directly -- rather, we enqueue a Task to run it. So it's not literally going to end up in an infinite cyclic call chain.

We should be safe from runaway behavior because:

(1) We only ever inject a new connection if there are requests queued -- and more specifically, if the # of requests queued > # of pending connections currently. (And other policy is true as well.)
(2) Any connection attempt will always result in one request being dequeued and processed (unless the queue is empty). In particular, if a connection attempt fails, this will result in a request getting failed.

So at the very least, the number of connection attempts will be bounded by the number of requests submitted. And as connections are reused for multiple requests, this will shrink the queue and further limit the number of connections created.

I do agree that there is a risk of subtle bugs here that could result in very problematic behavior, but I'm not sure what to do here beyond what I've already done -- try to keep the model as simple as possible, and assert where appropriate that unexpected things are not happening. Suggestions welcome.